cmd/testwrapper: eats race failures?

On the race builders, we've been seeing:

```
ok	tailscale.com/wgengine/wgint	(cached)
ok	tailscale.com/wgengine/wglog	(cached)
?	tailscale.com/wif [skipped/no tests] 
ok	tailscale.com/words	(cached)
ok	tailscale.com/wgengine/netstack	2.349
FAIL
FAIL	tailscale.com/wgengine/magicsock	10.363
ok	tailscale.com/tstest/integration	59.356
```

... with no failure output.

Is testwrapper eating failure output?

/cc @tailscale/tooling @icio @jwhited

I had Claude find all the interesting cases of the race failures and where they can land in the output stream that test2json has to then parse.

Its report:

---

# Go race output: test binary raw output vs `go tool test2json`

A survey of how the Go race detector's output is attributed to tests
(or not) under different test layouts. For each scenario, we have:

1. The test source.
2. The **raw** stdout+stderr of the test binary, captured by running
   `./pkg.test -test.v` directly (no `go test` wrapper).
3. The **JSON** output of feeding that exact raw output to
   `go tool test2json -p <pkg>`.

This is the same input/output relationship that `go test -json` uses
internally (it spawns the test binary with `-test.v=true` and pipes
its combined output through `test2json`). One additional thing
`go test` itself does, which `test2json` does not, is append a
`FAIL\t<pkg>\t<time>` line and a corresponding `fail` action event
when the binary exits non-zero — this matters for scenario E below,
where the binary prints `PASS` and *then* the race detector fires.

The key thing to take away is that **`test2json` attribution is
positional**: it tracks a single `c.testName` variable which is set
by `=== RUN <name>` lines, kept across `--- PASS/FAIL/SKIP: <name>`
lines, redirected by `=== NAME <name>` lines, and reset to `""` when
a pending PASS/FAIL is flushed. Any output line that isn't a
testing-framework directive gets stamped with whatever `c.testName`
currently is. The race detector prints from C (TSan), with no idea
which Go test "owns" the racing memory access — its output just
lands wherever the cursor happens to be in the test binary's
output stream.

## Why the race detector's view differs from `testing.T`'s view

There are two related but distinct things going on when `-race` is
enabled:

1. **The race detector (TSan in the runtime)** prints
   `WARNING: DATA RACE …` to the test binary's stderr the moment it
   observes the racy access. This text becomes test-binary output
   like anything else, and `test2json` attributes it by position.

2. **`testing.checkRaces`** runs at the end of each test in
   `tRunner`. It compares `race.Errors()` against
   `lastRaceErrors` (set by `resetRaces` at test start). If the
   counter incremented during the test, it calls
   `c.Errorf("race detected during execution of test")` — which is
   what marks the test as failed and emits the `testing.go:1712:`
   line and `--- FAIL: <name>` line.

These two are *not* synchronized. A race that fires between tests,
or after a test's `checkRaces` ran, will print a `WARNING: DATA RACE`
header but won't trip any test's `checkRaces` — so the test that
happens to be `c.testName` at the moment of printing will swallow the
race report in its log buffer even though it remains marked PASS.

---

## Scenario A: baseline (no race)

Sanity check. Two trivial passing tests. Shows the standard mapping
from test framework output lines to `test2json` events.

### Source

```go
package a

import "testing"

func TestOne(t *testing.T) {
	t.Log("hello from one")
}

func TestTwo(t *testing.T) {
}
```

### Raw output

```
=== RUN   TestOne
    a_test.go:6: hello from one
--- PASS: TestOne (0.00s)
=== RUN   TestTwo
--- PASS: TestTwo (0.00s)
PASS
```

### JSON output

```
{"Action":"start","Package":"A_plain"}
{"Action":"run","Package":"A_plain","Test":"TestOne"}
{"Action":"output","Package":"A_plain","Test":"TestOne","Output":"=== RUN   TestOne\n"}
{"Action":"output","Package":"A_plain","Test":"TestOne","Output":"    a_test.go:6: hello from one\n"}
{"Action":"output","Package":"A_plain","Test":"TestOne","Output":"--- PASS: TestOne (0.00s)\n"}
{"Action":"pass","Package":"A_plain","Test":"TestOne"}
{"Action":"run","Package":"A_plain","Test":"TestTwo"}
{"Action":"output","Package":"A_plain","Test":"TestTwo","Output":"=== RUN   TestTwo\n"}
{"Action":"output","Package":"A_plain","Test":"TestTwo","Output":"--- PASS: TestTwo (0.00s)\n"}
{"Action":"pass","Package":"A_plain","Test":"TestTwo"}
{"Action":"output","Package":"A_plain","Output":"PASS\n"}
{"Action":"pass","Package":"A_plain"}
```

Notes:

- Each `=== RUN <name>` line produces a `run` action event and an
  `output` event for the line itself; the `output` event carries the
  test name because `c.testName` is set as the line is parsed.
- Each `--- PASS: <name>` line is *queued* in `c.report`; the
  corresponding `pass` action event is emitted later when something
  flushes the report — either the next `=== RUN/PAUSE/CONT` line or
  `test2json.Close()` at end of input. In this trace you can see the
  `pass TestOne` event appearing *before* `run TestTwo` because the
  next `=== RUN` line triggered the flush.
- The bare `PASS` line at the end is the test binary's
  `testing.M.report()` summary. It is *not* attributed to any test
  (`c.testName` is `""` at this point because flushing reset it) and
  also sets `c.result = "pass"`, which is what `test2json` uses as
  the final package `pass`/`fail` action emitted from
  `test2json.Close()`.

---

## Scenario B: race during a single test's body

The race is synchronously produced by code that runs inside
`TestRace`. `checkRaces` catches it cleanly and the test is marked
failed.

### Source

```go
package b

import (
	"sync"
	"testing"
)

var counter int

func TestRace(t *testing.T) {
	var wg sync.WaitGroup
	wg.Add(2)
	go func() { defer wg.Done(); counter++ }()
	go func() { defer wg.Done(); counter++ }()
	wg.Wait()
}
```

### Raw output

```
=== RUN   TestRace
==================
WARNING: DATA RACE
Read at 0x0000007f12a8 by goroutine 8:
  racesurvey/B_inbody.TestRace.func1()
      /tmp/racesurvey/B_inbody/b_test.go:13 +0x74

Previous write at 0x0000007f12a8 by goroutine 9:
  racesurvey/B_inbody.TestRace.func2()
      /tmp/racesurvey/B_inbody/b_test.go:14 +0x8c

Goroutine 8 (running) created at:
  racesurvey/B_inbody.TestRace()
      /tmp/racesurvey/B_inbody/b_test.go:13 +0xbe
  testing.tRunner()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38

Goroutine 9 (finished) created at:
  racesurvey/B_inbody.TestRace()
      /tmp/racesurvey/B_inbody/b_test.go:14 +0x124
  testing.tRunner()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
==================
    testing.go:1712: race detected during execution of test
--- FAIL: TestRace (0.00s)
FAIL
```

### JSON output

```
{"Action":"start","Package":"B_inbody"}
{"Action":"run","Package":"B_inbody","Test":"TestRace"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"=== RUN   TestRace\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"==================\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"WARNING: DATA RACE\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"Read at 0x0000007f12a8 by goroutine 8:\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"  racesurvey/B_inbody.TestRace.func1()\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"      /tmp/racesurvey/B_inbody/b_test.go:13 +0x74\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"Previous write at 0x0000007f12a8 by goroutine 9:\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"  racesurvey/B_inbody.TestRace.func2()\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"      /tmp/racesurvey/B_inbody/b_test.go:14 +0x8c\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"Goroutine 8 (running) created at:\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"  racesurvey/B_inbody.TestRace()\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"      /tmp/racesurvey/B_inbody/b_test.go:13 +0xbe\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"  testing.tRunner()\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"  testing.(*T).Run.gowrap1()\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"Goroutine 9 (finished) created at:\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"  racesurvey/B_inbody.TestRace()\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"      /tmp/racesurvey/B_inbody/b_test.go:14 +0x124\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"  testing.tRunner()\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"  testing.(*T).Run.gowrap1()\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"==================\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"    testing.go:1712: race detected during execution of test\n"}
{"Action":"output","Package":"B_inbody","Test":"TestRace","Output":"--- FAIL: TestRace (0.00s)\n"}
{"Action":"fail","Package":"B_inbody","Test":"TestRace"}
{"Action":"output","Package":"B_inbody","Output":"FAIL\n"}
{"Action":"fail","Package":"B_inbody"}
```

Notes:

- Everything between `=== RUN TestRace` and `--- FAIL: TestRace` is
  attributed to TestRace, including the entire race report and the
  `race detected during execution of test` line.
- `testwrapper` (or any consumer that flushes logs of failed tests)
  will surface the full race report. This is the well-behaved case.

---

## Scenario C: spawn-then-wait (the bug scenario)

`TestSpawn` launches two goroutines that race on a global, then
returns. The goroutines outlive the test. `TestWait` then blocks on
the `WaitGroup` they signal. Where the race report ends up is
**purely a function of timing** — specifically, where the cursor in
the test binary's output stream is at the instant TSan decides to
print.

### Source

```go
package c

import (
	"sync"
	"testing"
)

var counter int
var wg sync.WaitGroup

func TestSpawn(t *testing.T) {
	wg.Add(2)
	go func() { defer wg.Done(); counter++ }()
	go func() { defer wg.Done(); counter++ }()
}

func TestWait(t *testing.T) {
	wg.Wait()
}
```

### Raw output (one observed run)

```
=== RUN   TestSpawn
--- PASS: TestSpawn (0.00s)
=== RUN   TestWait
==================
WARNING: DATA RACE
Read at 0x0000007f12a8 by goroutine 9:
  racesurvey/C_spawnwait.TestSpawn.func2()
      /tmp/racesurvey/C_spawnwait/c_test.go:14 +0x70

Previous write at 0x0000007f12a8 by goroutine 8:
  racesurvey/C_spawnwait.TestSpawn.func1()
      /tmp/racesurvey/C_spawnwait/c_test.go:13 +0x88

Goroutine 9 (running) created at:
  racesurvey/C_spawnwait.TestSpawn()
      /tmp/racesurvey/C_spawnwait/c_test.go:14 +0x44
  testing.tRunner()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38

Goroutine 8 (finished) created at:
  racesurvey/C_spawnwait.TestSpawn()
      /tmp/racesurvey/C_spawnwait/c_test.go:13 +0x34
  testing.tRunner()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
==================
--- PASS: TestWait (0.00s)
FAIL
```

### JSON output (from that same run)

```
{"Action":"start","Package":"C_spawnwait"}
{"Action":"run","Package":"C_spawnwait","Test":"TestSpawn"}
{"Action":"output","Package":"C_spawnwait","Test":"TestSpawn","Output":"=== RUN   TestSpawn\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestSpawn","Output":"--- PASS: TestSpawn (0.00s)\n"}
{"Action":"pass","Package":"C_spawnwait","Test":"TestSpawn"}
{"Action":"run","Package":"C_spawnwait","Test":"TestWait"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"=== RUN   TestWait\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"==================\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"WARNING: DATA RACE\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"Read at 0x0000007f12a8 by goroutine 9:\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"  racesurvey/C_spawnwait.TestSpawn.func2()\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"      /tmp/racesurvey/C_spawnwait/c_test.go:14 +0x70\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"Previous write at 0x0000007f12a8 by goroutine 8:\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"  racesurvey/C_spawnwait.TestSpawn.func1()\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"      /tmp/racesurvey/C_spawnwait/c_test.go:13 +0x88\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"Goroutine 9 (running) created at:\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"  racesurvey/C_spawnwait.TestSpawn()\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"      /tmp/racesurvey/C_spawnwait/c_test.go:14 +0x44\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"  testing.tRunner()\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"  testing.(*T).Run.gowrap1()\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"Goroutine 8 (finished) created at:\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"  racesurvey/C_spawnwait.TestSpawn()\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"      /tmp/racesurvey/C_spawnwait/c_test.go:13 +0x34\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"  testing.tRunner()\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"  testing.(*T).Run.gowrap1()\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"==================\n"}
{"Action":"output","Package":"C_spawnwait","Test":"TestWait","Output":"--- PASS: TestWait (0.00s)\n"}
{"Action":"pass","Package":"C_spawnwait","Test":"TestWait"}
{"Action":"output","Package":"C_spawnwait","Output":"FAIL\n"}
{"Action":"fail","Package":"C_spawnwait"}
```

Notes:

- In this run TSan's `WARNING: DATA RACE …` was printed *after*
  `=== RUN TestWait`, so `c.testName` was TestWait and the whole race
  report got attributed to TestWait.
- Critically, neither test was marked failed by the framework:
  - `TestSpawn`'s `checkRaces` ran before the race actually happened
    (the goroutines hadn't executed yet), so its
    `race.Errors()` delta was zero.
  - `TestWait`'s `checkRaces` saw `race.Errors() == lastRaceErrors`
    too, because the counter was already at the new value when
    `resetRaces` snapshotted it at TestWait start.
  - Both tests emit `--- PASS:`. Only the package as a whole exits
    nonzero, producing the bare `FAIL` line and the package-level
    `fail` action.
- A different run can put the WARNING before `--- PASS: TestSpawn`
  (race fires during `TestSpawn`'s body or its post-body window),
  attributing the race report to TestSpawn instead. Same shape of
  bug.
- This is the case `testwrapper` was silently dropping: the test
  with the `WARNING: DATA RACE` text in its log buffer is marked
  `pass`, so `testwrapper`'s default "only flush logs of failed
  tests" rule discards the race report.

---

## Scenario D: race forced into the next test's body

Same shape as C, but with a channel signal to deliberately delay the
racing goroutines until `TestSleep`'s body is running. This is the
*good* version of the spawn-then-wait pattern from the framework's
perspective: the increment to `race.Errors()` happens between
`TestSleep`'s `resetRaces` and its `checkRaces`, so `checkRaces`
catches it.

### Source

```go
package d

import (
	"sync"
	"testing"
	"time"
)

var counter int
var wg sync.WaitGroup
var trigger = make(chan struct{})

func TestA(t *testing.T) {
	wg.Add(2)
	go func() { defer wg.Done(); <-trigger; counter++ }()
	go func() { defer wg.Done(); <-trigger; counter++ }()
}

func TestSleep(t *testing.T) {
	close(trigger)
	// Sleep long enough that the goroutines race during this sleep,
	// but TestSleep itself doesn't write to counter.
	time.Sleep(50 * time.Millisecond)
	wg.Wait()
}
```

### Raw output

```
=== RUN   TestA
--- PASS: TestA (0.00s)
=== RUN   TestSleep
==================
WARNING: DATA RACE
Read at 0x0000007f12a8 by goroutine 9:
  racesurvey/D_postterminal.TestA.func2()
      /tmp/racesurvey/D_postterminal/d_test.go:16 +0x8a

Previous write at 0x0000007f12a8 by goroutine 8:
  racesurvey/D_postterminal.TestA.func1()
      /tmp/racesurvey/D_postterminal/d_test.go:15 +0xa4

Goroutine 9 (running) created at:
  racesurvey/D_postterminal.TestA()
      /tmp/racesurvey/D_postterminal/d_test.go:16 +0x44
  testing.tRunner()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38

Goroutine 8 (finished) created at:
  racesurvey/D_postterminal.TestA()
      /tmp/racesurvey/D_postterminal/d_test.go:15 +0x34
  testing.tRunner()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
==================
    testing.go:1712: race detected during execution of test
--- FAIL: TestSleep (0.05s)
FAIL
```

### JSON output

```
{"Action":"start","Package":"D_postterminal"}
{"Action":"run","Package":"D_postterminal","Test":"TestA"}
{"Action":"output","Package":"D_postterminal","Test":"TestA","Output":"=== RUN   TestA\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestA","Output":"--- PASS: TestA (0.00s)\n"}
{"Action":"pass","Package":"D_postterminal","Test":"TestA"}
{"Action":"run","Package":"D_postterminal","Test":"TestSleep"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"=== RUN   TestSleep\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"==================\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"WARNING: DATA RACE\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"Read at 0x0000007f12a8 by goroutine 9:\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"  racesurvey/D_postterminal.TestA.func2()\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"      /tmp/racesurvey/D_postterminal/d_test.go:16 +0x8a\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"Previous write at 0x0000007f12a8 by goroutine 8:\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"  racesurvey/D_postterminal.TestA.func1()\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"      /tmp/racesurvey/D_postterminal/d_test.go:15 +0xa4\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"Goroutine 9 (running) created at:\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"  racesurvey/D_postterminal.TestA()\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"      /tmp/racesurvey/D_postterminal/d_test.go:16 +0x44\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"  testing.tRunner()\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"  testing.(*T).Run.gowrap1()\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"Goroutine 8 (finished) created at:\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"  racesurvey/D_postterminal.TestA()\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"      /tmp/racesurvey/D_postterminal/d_test.go:15 +0x34\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"  testing.tRunner()\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"  testing.(*T).Run.gowrap1()\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"==================\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"    testing.go:1712: race detected during execution of test\n"}
{"Action":"output","Package":"D_postterminal","Test":"TestSleep","Output":"--- FAIL: TestSleep (0.05s)\n"}
{"Action":"fail","Package":"D_postterminal","Test":"TestSleep"}
{"Action":"output","Package":"D_postterminal","Output":"FAIL\n"}
{"Action":"fail","Package":"D_postterminal"}
```

Notes:

- The interesting contrast with C is just the addition of
  `testing.go:1712: race detected during execution of test`
  followed by `--- FAIL: TestSleep`. The race report itself looks
  identical, but here the framework's `checkRaces` saw the counter
  delta and called `t.Errorf`, producing the failure markers.
- Both C and D have `WARNING: DATA RACE` attributed to a single test
  in the JSON; the *only* observable difference is whether that test
  ends in PASS or FAIL.

---

## Scenario E: race in TestMain after `m.Run()`

The race-producing goroutines are spawned by `TestMain` *after*
`m.Run()` returns. By that point the testing framework has already
printed its final `PASS\n` summary, so `c.testName == ""` in
`test2json` and the entire race report is attributed at the
**package level** (no `Test` field on the output events).

### Source

```go
package e

import (
	"sync"
	"testing"
)

var counter int

func TestPass(t *testing.T) {
}

func TestMain(m *testing.M) {
	code := m.Run()
	var wg sync.WaitGroup
	wg.Add(2)
	go func() { defer wg.Done(); counter++ }()
	go func() { defer wg.Done(); counter++ }()
	wg.Wait()
	_ = code
}
```

### Raw output

```
=== RUN   TestPass
--- PASS: TestPass (0.00s)
PASS
==================
WARNING: DATA RACE
Read at 0x0000007f52a8 by goroutine 9:
  racesurvey/E_testmain.TestMain.func2()
      /tmp/racesurvey/E_testmain/e_test.go:18 +0x74

Previous write at 0x0000007f52a8 by goroutine 8:
  racesurvey/E_testmain.TestMain.func1()
      /tmp/racesurvey/E_testmain/e_test.go:17 +0x8c

Goroutine 9 (running) created at:
  racesurvey/E_testmain.TestMain()
      /tmp/racesurvey/E_testmain/e_test.go:18 +0x139
  main.main()
      _testmain.go:48 +0x171

Goroutine 8 (finished) created at:
  racesurvey/E_testmain.TestMain()
      /tmp/racesurvey/E_testmain/e_test.go:17 +0xcc
  main.main()
      _testmain.go:48 +0x171
==================
Found 1 data race(s)
```

### JSON output

```
{"Action":"start","Package":"E_testmain"}
{"Action":"run","Package":"E_testmain","Test":"TestPass"}
{"Action":"output","Package":"E_testmain","Test":"TestPass","Output":"=== RUN   TestPass\n"}
{"Action":"output","Package":"E_testmain","Test":"TestPass","Output":"--- PASS: TestPass (0.00s)\n"}
{"Action":"pass","Package":"E_testmain","Test":"TestPass"}
{"Action":"output","Package":"E_testmain","Output":"PASS\n"}
{"Action":"output","Package":"E_testmain","Output":"==================\n"}
{"Action":"output","Package":"E_testmain","Output":"WARNING: DATA RACE\n"}
{"Action":"output","Package":"E_testmain","Output":"Read at 0x0000007f52a8 by goroutine 9:\n"}
{"Action":"output","Package":"E_testmain","Output":"  racesurvey/E_testmain.TestMain.func2()\n"}
{"Action":"output","Package":"E_testmain","Output":"      /tmp/racesurvey/E_testmain/e_test.go:18 +0x74\n"}
{"Action":"output","Package":"E_testmain","Output":"\n"}
{"Action":"output","Package":"E_testmain","Output":"Previous write at 0x0000007f52a8 by goroutine 8:\n"}
{"Action":"output","Package":"E_testmain","Output":"  racesurvey/E_testmain.TestMain.func1()\n"}
{"Action":"output","Package":"E_testmain","Output":"      /tmp/racesurvey/E_testmain/e_test.go:17 +0x8c\n"}
{"Action":"output","Package":"E_testmain","Output":"\n"}
{"Action":"output","Package":"E_testmain","Output":"Goroutine 9 (running) created at:\n"}
{"Action":"output","Package":"E_testmain","Output":"  racesurvey/E_testmain.TestMain()\n"}
{"Action":"output","Package":"E_testmain","Output":"      /tmp/racesurvey/E_testmain/e_test.go:18 +0x139\n"}
{"Action":"output","Package":"E_testmain","Output":"  main.main()\n"}
{"Action":"output","Package":"E_testmain","Output":"      _testmain.go:48 +0x171\n"}
{"Action":"output","Package":"E_testmain","Output":"\n"}
{"Action":"output","Package":"E_testmain","Output":"Goroutine 8 (finished) created at:\n"}
{"Action":"output","Package":"E_testmain","Output":"  racesurvey/E_testmain.TestMain()\n"}
{"Action":"output","Package":"E_testmain","Output":"      /tmp/racesurvey/E_testmain/e_test.go:17 +0xcc\n"}
{"Action":"output","Package":"E_testmain","Output":"  main.main()\n"}
{"Action":"output","Package":"E_testmain","Output":"      _testmain.go:48 +0x171\n"}
{"Action":"output","Package":"E_testmain","Output":"==================\n"}
{"Action":"output","Package":"E_testmain","Output":"Found 1 data race(s)\n"}
{"Action":"pass","Package":"E_testmain"}
```

Notes:

- All race output events have no `Test` field — they are
  package-level.
- The final `pass` action on the package is **wrong** but it's what
  `test2json` emits in isolation. `c.result` was set to `"pass"` when
  the bare `PASS\n` line was processed earlier, and `test2json` has
  no way to know the process subsequently exited with code 66.
- When `go test` is the wrapper, it observes the non-zero exit and
  appends its own `FAIL\t<pkg>\t<time>` output line plus a `fail`
  action event, overriding what `test2json.Close()` would have
  emitted. So real consumers (testwrapper, IDEs) see the package as
  failed and the race report in package-level logs — fine.

---

## Scenario F: parallel tests racing on shared state

Two `t.Parallel()` tests both increment the same global. The race
report is **split across two tests**: the `WARNING: DATA RACE` header
plus the full stack trace gets attributed to one test (which still
PASSes), then a `=== NAME` directive redirects attribution to the
other test for the trailing `race detected during execution of test`
line and `--- FAIL` line.

### Source

```go
package f

import "testing"

var counter int

func TestParA(t *testing.T) {
	t.Parallel()
	for i := 0; i < 100; i++ {
		counter++
	}
}

func TestParB(t *testing.T) {
	t.Parallel()
	for i := 0; i < 100; i++ {
		counter++
	}
}
```

### Raw output

```
=== RUN   TestParA
=== PAUSE TestParA
=== RUN   TestParB
=== PAUSE TestParB
=== CONT  TestParA
=== CONT  TestParB
--- PASS: TestParA (0.00s)
==================
WARNING: DATA RACE
Read at 0x0000007f02b0 by goroutine 8:
  racesurvey/F_parallel.TestParB()
      /tmp/racesurvey/F_parallel/f_test.go:17 +0x3b
  testing.tRunner()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38

Previous write at 0x0000007f02b0 by goroutine 7:
  racesurvey/F_parallel.TestParA()
      /tmp/racesurvey/F_parallel/f_test.go:10 +0x53
  testing.tRunner()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38

Goroutine 8 (running) created at:
  testing.(*T).Run()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0xb12
  testing.runTests.func1()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2585 +0x84
  testing.tRunner()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
  testing.runTests()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2583 +0x9e9
  testing.(*M).Run()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2443 +0xf4b
  main.main()
      _testmain.go:48 +0x164

Goroutine 7 (running) created at:
  testing.(*T).Run()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0xb12
  testing.runTests.func1()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2585 +0x84
  testing.tRunner()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
  testing.runTests()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2583 +0x9e9
  testing.(*M).Run()
      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2443 +0xf4b
  main.main()
      _testmain.go:48 +0x164
==================
=== NAME  TestParB
    testing.go:1712: race detected during execution of test
--- FAIL: TestParB (0.00s)
FAIL
```

### JSON output

```
{"Action":"start","Package":"F_parallel"}
{"Action":"run","Package":"F_parallel","Test":"TestParA"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"=== RUN   TestParA\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"=== PAUSE TestParA\n"}
{"Action":"pause","Package":"F_parallel","Test":"TestParA"}
{"Action":"run","Package":"F_parallel","Test":"TestParB"}
{"Action":"output","Package":"F_parallel","Test":"TestParB","Output":"=== RUN   TestParB\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParB","Output":"=== PAUSE TestParB\n"}
{"Action":"pause","Package":"F_parallel","Test":"TestParB"}
{"Action":"cont","Package":"F_parallel","Test":"TestParA"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"=== CONT  TestParA\n"}
{"Action":"cont","Package":"F_parallel","Test":"TestParB"}
{"Action":"output","Package":"F_parallel","Test":"TestParB","Output":"=== CONT  TestParB\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"--- PASS: TestParA (0.00s)\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"==================\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"WARNING: DATA RACE\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"Read at 0x0000007f02b0 by goroutine 8:\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  racesurvey/F_parallel.TestParB()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /tmp/racesurvey/F_parallel/f_test.go:17 +0x3b\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.tRunner()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.(*T).Run.gowrap1()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"Previous write at 0x0000007f02b0 by goroutine 7:\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  racesurvey/F_parallel.TestParA()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /tmp/racesurvey/F_parallel/f_test.go:10 +0x53\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.tRunner()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.(*T).Run.gowrap1()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"Goroutine 8 (running) created at:\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.(*T).Run()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0xb12\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.runTests.func1()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2585 +0x84\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.tRunner()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.runTests()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2583 +0x9e9\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.(*M).Run()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2443 +0xf4b\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  main.main()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      _testmain.go:48 +0x164\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"Goroutine 7 (running) created at:\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.(*T).Run()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0xb12\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.runTests.func1()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2585 +0x84\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.tRunner()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.runTests()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2583 +0x9e9\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  testing.(*M).Run()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2443 +0xf4b\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"  main.main()\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"      _testmain.go:48 +0x164\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParA","Output":"==================\n"}
{"Action":"pass","Package":"F_parallel","Test":"TestParA"}
{"Action":"output","Package":"F_parallel","Test":"TestParB","Output":"    testing.go:1712: race detected during execution of test\n"}
{"Action":"output","Package":"F_parallel","Test":"TestParB","Output":"--- FAIL: TestParB (0.00s)\n"}
{"Action":"fail","Package":"F_parallel","Test":"TestParB"}
{"Action":"output","Package":"F_parallel","Output":"FAIL\n"}
{"Action":"fail","Package":"F_parallel"}
```

Notes:

- The raw output includes the line `=== NAME  TestParB`. This is a
  directive the testing framework emits (via `chattyPrinter`) for
  the explicit purpose of telling `test2json` "switch attribution to
  TestParB now." It produces no JSON event of its own — it just
  silently sets `c.testName = "TestParB"` so subsequent lines get
  the new attribution.
- As a result, the JSON shows the race report attributed to
  TestParA (which is `pass`), and the trailing
  `race detected during execution of test` + `--- FAIL` lines
  attributed to TestParB (which is `fail`).
- A naive consumer that only flushes the failed test's logs sees
  the trailing `race detected during execution of test` line in
  TestParB's logs but has no idea *what* the race was, because the
  actual `WARNING: DATA RACE` block is in TestParA's (passing)
  buffer.

---

## Summary

| Scenario | Where `WARNING: DATA RACE` lands | That test's outcome | Visible to default `testwrapper`? |
|---|---|---|---|
| A baseline | n/a | n/a | n/a |
| B in-body | the test | fail | yes |
| C spawn/wait (timing-A) | TestSpawn | pass | **no** |
| C spawn/wait (timing-B) | TestWait | pass | **no** |
| C spawn/wait (timing-C) | TestWait | fail | yes |
| D delayed | TestSleep | fail | yes |
| E TestMain post-Run | pkg level | pkg fail | yes |
| F parallel | passing test (`WARNING` + stack) **and** failing test (`race detected` + `FAIL`) | pass + fail (split) | partial — only the trailing half |

The `testwrapper` bug surfaces in scenarios C-with-unlucky-timing
and F: the `WARNING: DATA RACE` text is sitting in a passing test's
log buffer that the wrapper would normally discard. Two viable
fixes:

1. **Promote on marker.** Scan test-attributed output for
   `WARNING: DATA RACE`. If we see it and the test ends in `pass` or
   `skip`, override the outcome to `fail` so the main loop flushes
   its logs through the usual code path.
2. **Fold passing logs on package fail.** When a package fails and
   no test in it was marked fail, attach all passing tests' log
   buffers to the package-level logs and flush. Broader (catches
   any anomalous output, not just races) but noisier.

--- a/cmd/testwrapper/testdata/A_baseline/raw.txt
+++ b/cmd/testwrapper/testdata/A_baseline/raw.txt
@@ -1,0 +1,6 @@
+=== RUN   TestOne
+    a_test.go:6: hello from one
+--- PASS: TestOne (0.00s)
+=== RUN   TestTwo
+--- PASS: TestTwo (0.00s)
+PASS

--- a/cmd/testwrapper/testdata/A_baseline/src.go
+++ b/cmd/testwrapper/testdata/A_baseline/src.go
@@ -1,0 +1,13 @@
+// Copyright (c) Tailscale Inc & contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+package baseline
+
+import "testing"
+
+func TestOne(t *testing.T) {
+	t.Log("hello from one")
+}
+
+func TestTwo(t *testing.T) {
+}

--- a/cmd/testwrapper/testdata/B_inbody/raw.txt
+++ b/cmd/testwrapper/testdata/B_inbody/raw.txt
@@ -1,0 +1,30 @@
+=== RUN   TestRace
+==================
+WARNING: DATA RACE
+Read at 0x0000007f12a8 by goroutine 8:
+  racesurvey/B_inbody.TestRace.func1()
+      /tmp/racesurvey/B_inbody/b_test.go:13 +0x74
+
+Previous write at 0x0000007f12a8 by goroutine 9:
+  racesurvey/B_inbody.TestRace.func2()
+      /tmp/racesurvey/B_inbody/b_test.go:14 +0x8c
+
+Goroutine 8 (running) created at:
+  racesurvey/B_inbody.TestRace()
+      /tmp/racesurvey/B_inbody/b_test.go:13 +0xbe
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.(*T).Run.gowrap1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
+
+Goroutine 9 (finished) created at:
+  racesurvey/B_inbody.TestRace()
+      /tmp/racesurvey/B_inbody/b_test.go:14 +0x124
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.(*T).Run.gowrap1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
+==================
+    testing.go:1712: race detected during execution of test
+--- FAIL: TestRace (0.00s)
+FAIL

--- a/cmd/testwrapper/testdata/B_inbody/src.go
+++ b/cmd/testwrapper/testdata/B_inbody/src.go
@@ -1,0 +1,19 @@
+// Copyright (c) Tailscale Inc & contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+package inbody
+
+import (
+	"sync"
+	"testing"
+)
+
+var counter int
+
+func TestRace(t *testing.T) {
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() { defer wg.Done(); counter++ }()
+	go func() { defer wg.Done(); counter++ }()
+	wg.Wait()
+}

--- a/cmd/testwrapper/testdata/C_spawnwait/caught.raw.txt
+++ b/cmd/testwrapper/testdata/C_spawnwait/caught.raw.txt
@@ -1,0 +1,32 @@
+=== RUN   TestSpawn
+==================
+WARNING: DATA RACE
+Read at 0x0000007f12a8 by goroutine 9:
+  racesurvey/C_spawnwait.TestSpawn.func2()
+      /tmp/racesurvey/C_spawnwait/c_test.go:14 +0x70
+
+Previous write at 0x0000007f12a8 by goroutine 8:
+  racesurvey/C_spawnwait.TestSpawn.func1()
+      /tmp/racesurvey/C_spawnwait/c_test.go:13 +0x88
+
+Goroutine 9 (running) created at:
+  racesurvey/C_spawnwait.TestSpawn()
+      /tmp/racesurvey/C_spawnwait/c_test.go:14 +0x44
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.(*T).Run.gowrap1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
+
+Goroutine 8 (finished) created at:
+  racesurvey/C_spawnwait.TestSpawn()
+      /tmp/racesurvey/C_spawnwait/c_test.go:13 +0x34
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.(*T).Run.gowrap1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
+==================
+    testing.go:1712: race detected during execution of test
+--- FAIL: TestSpawn (0.00s)
+=== RUN   TestWait
+--- PASS: TestWait (0.00s)
+FAIL

--- a/cmd/testwrapper/testdata/C_spawnwait/pass.raw.txt
+++ b/cmd/testwrapper/testdata/C_spawnwait/pass.raw.txt
@@ -1,0 +1,31 @@
+=== RUN   TestSpawn
+--- PASS: TestSpawn (0.00s)
+=== RUN   TestWait
+==================
+WARNING: DATA RACE
+Read at 0x0000007f12a8 by goroutine 8:
+  racesurvey/C_spawnwait.TestSpawn.func1()
+      /tmp/racesurvey/C_spawnwait/c_test.go:13 +0x70
+
+Previous write at 0x0000007f12a8 by goroutine 9:
+  racesurvey/C_spawnwait.TestSpawn.func2()
+      /tmp/racesurvey/C_spawnwait/c_test.go:14 +0x88
+
+Goroutine 8 (running) created at:
+  racesurvey/C_spawnwait.TestSpawn()
+      /tmp/racesurvey/C_spawnwait/c_test.go:13 +0x34
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.(*T).Run.gowrap1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
+
+Goroutine 9 (finished) created at:
+  racesurvey/C_spawnwait.TestSpawn()
+      /tmp/racesurvey/C_spawnwait/c_test.go:14 +0x44
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.(*T).Run.gowrap1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
+==================
+--- PASS: TestWait (0.00s)
+FAIL

--- a/cmd/testwrapper/testdata/C_spawnwait/src.go
+++ b/cmd/testwrapper/testdata/C_spawnwait/src.go
@@ -1,0 +1,22 @@
+// Copyright (c) Tailscale Inc & contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+package spawnwait
+
+import (
+	"sync"
+	"testing"
+)
+
+var counter int
+var wg sync.WaitGroup
+
+func TestSpawn(t *testing.T) {
+	wg.Add(2)
+	go func() { defer wg.Done(); counter++ }()
+	go func() { defer wg.Done(); counter++ }()
+}
+
+func TestWait(t *testing.T) {
+	wg.Wait()
+}

--- a/cmd/testwrapper/testdata/D_delayed/raw.txt
+++ b/cmd/testwrapper/testdata/D_delayed/raw.txt
@@ -1,0 +1,32 @@
+=== RUN   TestA
+--- PASS: TestA (0.00s)
+=== RUN   TestSleep
+==================
+WARNING: DATA RACE
+Read at 0x0000007f12a8 by goroutine 9:
+  racesurvey/D_postterminal.TestA.func2()
+      /tmp/racesurvey/D_postterminal/d_test.go:16 +0x8a
+
+Previous write at 0x0000007f12a8 by goroutine 8:
+  racesurvey/D_postterminal.TestA.func1()
+      /tmp/racesurvey/D_postterminal/d_test.go:15 +0xa4
+
+Goroutine 9 (running) created at:
+  racesurvey/D_postterminal.TestA()
+      /tmp/racesurvey/D_postterminal/d_test.go:16 +0x44
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.(*T).Run.gowrap1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
+
+Goroutine 8 (finished) created at:
+  racesurvey/D_postterminal.TestA()
+      /tmp/racesurvey/D_postterminal/d_test.go:15 +0x34
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.(*T).Run.gowrap1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
+==================
+    testing.go:1712: race detected during execution of test
+--- FAIL: TestSleep (0.05s)
+FAIL

--- a/cmd/testwrapper/testdata/D_delayed/src.go
+++ b/cmd/testwrapper/testdata/D_delayed/src.go
@@ -1,0 +1,28 @@
+// Copyright (c) Tailscale Inc & contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+package delayed
+
+import (
+	"sync"
+	"testing"
+	"time"
+)
+
+var counter int
+var wg sync.WaitGroup
+var trigger = make(chan struct{})
+
+func TestA(t *testing.T) {
+	wg.Add(2)
+	go func() { defer wg.Done(); <-trigger; counter++ }()
+	go func() { defer wg.Done(); <-trigger; counter++ }()
+}
+
+func TestSleep(t *testing.T) {
+	close(trigger)
+	// Sleep long enough that the goroutines race during this sleep,
+	// but TestSleep itself doesn't write to counter.
+	time.Sleep(50 * time.Millisecond)
+	wg.Wait()
+}

--- a/cmd/testwrapper/testdata/E_testmain/raw.txt
+++ b/cmd/testwrapper/testdata/E_testmain/raw.txt
@@ -1,0 +1,26 @@
+=== RUN   TestPass
+--- PASS: TestPass (0.00s)
+PASS
+==================
+WARNING: DATA RACE
+Read at 0x0000007f52a8 by goroutine 9:
+  racesurvey/E_testmain.TestMain.func2()
+      /tmp/racesurvey/E_testmain/e_test.go:18 +0x74
+
+Previous write at 0x0000007f52a8 by goroutine 8:
+  racesurvey/E_testmain.TestMain.func1()
+      /tmp/racesurvey/E_testmain/e_test.go:17 +0x8c
+
+Goroutine 9 (running) created at:
+  racesurvey/E_testmain.TestMain()
+      /tmp/racesurvey/E_testmain/e_test.go:18 +0x139
+  main.main()
+      _testmain.go:48 +0x171
+
+Goroutine 8 (finished) created at:
+  racesurvey/E_testmain.TestMain()
+      /tmp/racesurvey/E_testmain/e_test.go:17 +0xcc
+  main.main()
+      _testmain.go:48 +0x171
+==================
+Found 1 data race(s)

--- a/cmd/testwrapper/testdata/E_testmain/src.go
+++ b/cmd/testwrapper/testdata/E_testmain/src.go
@@ -1,0 +1,24 @@
+// Copyright (c) Tailscale Inc & contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+package testmain
+
+import (
+	"sync"
+	"testing"
+)
+
+var counter int
+
+func TestPass(t *testing.T) {
+}
+
+func TestMain(m *testing.M) {
+	code := m.Run()
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() { defer wg.Done(); counter++ }()
+	go func() { defer wg.Done(); counter++ }()
+	wg.Wait()
+	_ = code
+}

--- a/cmd/testwrapper/testdata/F_parallel/split.raw.txt
+++ b/cmd/testwrapper/testdata/F_parallel/split.raw.txt
@@ -1,0 +1,57 @@
+=== RUN   TestParA
+=== PAUSE TestParA
+=== RUN   TestParB
+=== PAUSE TestParB
+=== CONT  TestParA
+=== CONT  TestParB
+--- PASS: TestParA (0.00s)
+==================
+WARNING: DATA RACE
+Read at 0x0000007f02b0 by goroutine 8:
+  racesurvey/F_parallel.TestParB()
+      /tmp/racesurvey/F_parallel/f_test.go:17 +0x3b
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.(*T).Run.gowrap1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
+
+Previous write at 0x0000007f02b0 by goroutine 7:
+  racesurvey/F_parallel.TestParA()
+      /tmp/racesurvey/F_parallel/f_test.go:10 +0x53
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.(*T).Run.gowrap1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0x38
+
+Goroutine 8 (running) created at:
+  testing.(*T).Run()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0xb12
+  testing.runTests.func1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2585 +0x84
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.runTests()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2583 +0x9e9
+  testing.(*M).Run()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2443 +0xf4b
+  main.main()
+      _testmain.go:48 +0x164
+
+Goroutine 7 (running) created at:
+  testing.(*T).Run()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2101 +0xb12
+  testing.runTests.func1()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2585 +0x84
+  testing.tRunner()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2036 +0x21c
+  testing.runTests()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2583 +0x9e9
+  testing.(*M).Run()
+      /home/ubuntu/sdk/go1.26.3/src/testing/testing.go:2443 +0xf4b
+  main.main()
+      _testmain.go:48 +0x164
+==================
+=== NAME  TestParB
+    testing.go:1712: race detected during execution of test
+--- FAIL: TestParB (0.00s)
+FAIL

--- a/cmd/testwrapper/testdata/F_parallel/src.go
+++ b/cmd/testwrapper/testdata/F_parallel/src.go
@@ -1,0 +1,22 @@
+// Copyright (c) Tailscale Inc & contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+package parallel
+
+import "testing"
+
+var counter int
+
+func TestParA(t *testing.T) {
+	t.Parallel()
+	for i := 0; i < 100; i++ {
+		counter++
+	}
+}
+
+func TestParB(t *testing.T) {
+	t.Parallel()
+	for i := 0; i < 100; i++ {
+		counter++
+	}
+}

--- a/cmd/testwrapper/testdata/README.md
+++ b/cmd/testwrapper/testdata/README.md
@@ -1,0 +1,133 @@
+# Race-output test corpus
+
+This directory is a corpus of captured Go test binary outputs that
+exercise the various ways the `-race` detector's `WARNING: DATA RACE`
+text can land relative to `=== RUN` / `--- PASS:` / `--- FAIL:` /
+`=== NAME` lines, and how `cmd/internal/test2json` attributes that
+output to tests.
+
+Each scenario subdirectory contains:
+
+- `src.go` — the Go source code that was compiled and run to produce
+  the captured output. Reproduce via
+  `go test -race -c -o /tmp/scenario.test ./<dir>/`.
+- `raw.txt` (or scenario-specific name) — the raw stdout+stderr of
+  the resulting test binary when run as `./scenario.test -test.v`.
+  This is the byte stream that `go test -json` feeds to
+  `go tool test2json` in production.
+
+`go test -json` adds two things on top of what `test2json` sees,
+which are NOT in these captures: a `FAIL\t<pkg>\t<time>` output line
+and a corresponding `fail` action event when the binary exits with a
+non-zero status (e.g. the race detector's exit code 66). Consumers
+that simulate the `go test -json` view from these files need to
+append those.
+
+To turn a `raw.txt` into the JSON `testwrapper` would consume:
+
+```
+go tool test2json -p <pkgname> < raw.txt
+```
+
+## Scenarios
+
+### A_baseline — no race
+
+Two trivial passing tests, no `-race` involvement at all. Useful as a
+sanity-check for parsers; demonstrates the standard mapping from
+`=== RUN`/`--- PASS:`/bare `PASS` lines to JSON events.
+
+### B_inbody — race during a test's own body
+
+A single test spawns racing goroutines and `wg.Wait()`s for them.
+The race fires while the test is still running, so
+`testing.checkRaces` sees `race.Errors()` increment between the
+test's start and end and the test is marked `--- FAIL:` with a
+`testing.go:1712: race detected during execution of test` line.
+
+This is the well-behaved case: the race output and the failure marker
+are both attributed to a single, clearly-failed test.
+
+### C_spawnwait — race in goroutines that outlive a test
+
+`TestSpawn` spawns racing goroutines and returns immediately;
+`TestWait` blocks until they finish. Where the race report lands is
+timing-dependent. Two captured variants:
+
+- `pass.raw.txt` — both `TestSpawn` and `TestWait` end up marked
+  `--- PASS:`, with the entire `WARNING: DATA RACE` report
+  attributed to `TestWait` (the test whose `=== RUN` line was the
+  most recent thing emitted when TSan printed). Neither test's
+  `checkRaces` saw `race.Errors()` increment within its window. The
+  binary still exits non-zero, producing only a bare trailing
+  `FAIL`. **This is the case where testwrapper currently drops the
+  race report on the floor.**
+
+- `caught.raw.txt` — `TestSpawn`'s `checkRaces` happened to fire
+  after the goroutines raced, so it emits the standard
+  `testing.go:1712:` + `--- FAIL: TestSpawn` markers. Race report is
+  attributed to `TestSpawn`, and the failure is attributed to
+  `TestSpawn` too. Well-behaved.
+
+### D_delayed — race delayed into the next test's body
+
+Like `C_spawnwait` but the spawned goroutines block on a channel that
+the second test closes, then sleeps. This forces the race counter
+delta to land between the second test's `resetRaces` and
+`checkRaces`, so the framework reliably attributes a failure to the
+sleeping test. Compare with `B_inbody`: same end shape, but the
+failure is attributed to a test that doesn't *contain* the racing
+code.
+
+### E_testmain — race in `TestMain` after `m.Run()`
+
+The race-producing goroutines are spawned by `TestMain` after
+`m.Run()` returns. By that point the framework has already printed
+its top-level `PASS` summary, so `test2json`'s `c.testName` is reset
+to `""` and the entire race report shows up at **package level** —
+all of its `output` events have an empty `Test` field. The bare
+`PASS` line earlier in the output also causes `test2json` to set
+`c.result = "pass"`, which means `go tool test2json` in isolation
+ends up emitting `{"Action":"pass","Package":"..."}` for the
+package; only `go test`'s own exit-code post-processing turns it
+into the visible package-level failure.
+
+### F_parallel — racing parallel tests
+
+Two `t.Parallel()` tests both increment the same global. The
+captured `split.raw.txt` shows the report **split across two tests**
+via a `=== NAME  TestParB` redirect line emitted by the framework:
+
+```
+--- PASS: TestParA (0.00s)
+==================
+WARNING: DATA RACE
+   ... full race report ...
+==================
+=== NAME  TestParB
+    testing.go:1712: race detected during execution of test
+--- FAIL: TestParB (0.00s)
+```
+
+`test2json` attributes the `WARNING: DATA RACE` block plus the full
+stack trace to `TestParA` (which passes), then sees the
+`=== NAME  TestParB` directive and switches attribution, so the
+trailing `race detected` line and `--- FAIL` end up under
+`TestParB` (which fails). A consumer that only flushes failed tests'
+logs sees the `race detected` line but loses the stack trace.
+
+## How these were captured
+
+```
+# For each scenario directory <name>/ (e.g. C_spawnwait):
+go test -race -c -o /tmp/<name>.test ./testdata/<name>/
+/tmp/<name>.test -test.v >./testdata/<name>/raw.txt 2>&1
+```
+
+For scenarios where the race detector's attribution is
+timing-dependent (C, F), the binary was run repeatedly and the
+interesting variants saved with distinct names.
+
+The captured outputs contain absolute file paths from the machine
+they were recorded on (e.g. `/tmp/racesurvey/...`). Tests that
+match against these files should not depend on those paths.

--- a/cmd/testwrapper/testwrapper.go
+++ b/cmd/testwrapper/testwrapper.go
@@ -30,6 +30,14 @@ import (
 
 const (
 	maxAttempts = 3
+
+	// raceDetectorMarkerLine is the first line of every Go race
+	// detector report, emitted at column 0. We look for it as a
+	// whole line (not as a substring) so that we don't false-fire
+	// on tests that legitimately print the same text indented in
+	// their own logs — for example, this package's own race tests,
+	// which exec a child testwrapper and dump its captured output.
+	raceDetectorMarkerLine = "WARNING: DATA RACE\n"
 )
 
 type testAttempt struct {
@@ -41,6 +49,10 @@ type testAttempt struct {
 	start, end    time.Time
 	isMarkedFlaky bool   // set if the test is marked as flaky
 	issueURL      string // set if the test is marked as flaky
+	// raceDetected is true on a per-test event if that test's output
+	// contained a race report, and true on a pkgFinished event if any
+	// test in the package -- or the package's own output -- did.
+	raceDetected bool
 
 	pkgFinished bool
 }
@@ -153,14 +165,46 @@ func runTests(ctx context.Context, attempt int, pt *packageTests, goTestArgs, te
 					outcome = "fail"
 				}
 				pkgTests[""].logs.WriteString(goOutput.Output)
+				// If a data race was detected anywhere in this
+				// package's output -- whether at the package level or
+				// attributed to a specific test -- consolidate all
+				// per-test logs into the package-level logs so the
+				// full race report is visible regardless of which
+				// test test2json happened to attribute it to. The
+				// pkgFinished testAttempt also carries raceDetected
+				// so the main loop can suppress flaky-test retries.
+				raceDetected := pkgTests[""].raceDetected
+				if !raceDetected {
+					for _, t := range pkgTests {
+						if t.raceDetected {
+							raceDetected = true
+							break
+						}
+					}
+				}
+				if raceDetected {
+					var ts []*testAttempt
+					for _, t := range pkgTests {
+						if t.testName != "" && t.logs.Len() > 0 {
+							ts = append(ts, t)
+						}
+					}
+					slices.SortFunc(ts, func(a, b *testAttempt) int {
+						return a.start.Compare(b.start)
+					})
+					for _, t := range ts {
+						pkgTests[""].logs.Write(t.logs.Bytes())
+					}
+				}
 				ch <- &testAttempt{
-					pkg:         goOutput.Package,
-					outcome:     outcome,
-					start:       pkgTests[""].start,
-					end:         goOutput.Time,
-					logs:        pkgTests[""].logs,
-					pkgFinished: true,
-					cached:      pkgCached[goOutput.Package],
+					pkg:          goOutput.Package,
+					outcome:      outcome,
+					start:        pkgTests[""].start,
+					end:          goOutput.Time,
+					logs:         pkgTests[""].logs,
+					pkgFinished:  true,
+					cached:       pkgCached[goOutput.Package],
+					raceDetected: raceDetected,
 				}
 			case "output":
 				// Capture all output from the package except for the final
@@ -168,6 +212,9 @@ func runTests(ctx context.Context, attempt int, pt *packageTests, goTestArgs, te
 				// printPkgOutcome will output a similar line
 				if !strings.HasPrefix(goOutput.Output, fmt.Sprintf("FAIL\t%s\t", goOutput.Package)) {
 					pkgTests[""].logs.WriteString(goOutput.Output)
+					if goOutput.Output == raceDetectorMarkerLine {
+						pkgTests[""].raceDetected = true
+					}
 				}
 			}
 
@@ -178,6 +225,9 @@ func runTests(ctx context.Context, attempt int, pt *packageTests, goTestArgs, te
 			testName = test
 			if goOutput.Action == "output" {
 				resultMap[pkg][testName].logs.WriteString(goOutput.Output)
+				if goOutput.Output == raceDetectorMarkerLine {
+					resultMap[pkg][testName].raceDetected = true
+				}
 			}
 			continue
 		}
@@ -200,6 +250,9 @@ func runTests(ctx context.Context, attempt int, pt *packageTests, goTestArgs, te
 				pkgTests[testName].issueURL = strings.TrimPrefix(suffix, ": ")
 			} else {
 				pkgTests[testName].logs.WriteString(goOutput.Output)
+				if goOutput.Output == raceDetectorMarkerLine {
+					pkgTests[testName].raceDetected = true
+				}
 			}
 		}
 	}
@@ -305,6 +358,16 @@ func main() {
 					tr.pkg = packages[0]
 				}
 				if tr.pkgFinished {
+					if tr.raceDetected {
+						// A data race is never something we want to
+						// paper over by retrying flaky tests in the
+						// package: the race indicates a real bug
+						// that may not even be in the failing test,
+						// and a retry could hide it. Discard any
+						// retry plans for this pkg and fail fast.
+						delete(toRetry, tr.pkg)
+						failed = true
+					}
 					if tr.outcome == "fail" && len(toRetry[tr.pkg]) == 0 {
 						// If a package fails and we don't have any tests to
 						// retry, then we should fail. This typically happens

--- a/cmd/testwrapper/testwrapper_test.go
+++ b/cmd/testwrapper/testwrapper_test.go
@@ -145,6 +145,118 @@ func TestAlwaysError(t *testing.T) {
 	}
 }
 
+// TestRaceSuppressesFlakyRetry verifies that detecting a data race
+// in a package's output stops testwrapper from retrying any flaky
+// test in that package. Races are too serious to paper over: the
+// flaky test might not even be the one whose code is racy, and a
+// retry without the racy goroutine could silently hide the bug.
+func TestRaceSuppressesFlakyRetry(t *testing.T) {
+	if runtime.GOOS != "linux" || runtime.GOARCH != "amd64" {
+		t.Skip("test requires the race detector, which needs linux/amd64")
+	}
+	t.Parallel()
+
+	testfile := filepath.Join(t.TempDir(), "raceretry_test.go")
+	code := []byte(`package raceretry_test
+
+import (
+	"sync"
+	"testing"
+	"tailscale.com/cmd/testwrapper/flakytest"
+)
+
+var counter int
+
+func TestRace(t *testing.T) {
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() { defer wg.Done(); counter++ }()
+	go func() { defer wg.Done(); counter++ }()
+	wg.Wait()
+}
+
+func TestFlaky(t *testing.T) {
+	flakytest.Mark(t, "https://github.com/tailscale/tailscale/issues/0")
+	t.Fatal("flaky test failing; would normally be retried")
+}
+`)
+	if err := os.WriteFile(testfile, code, 0o644); err != nil {
+		t.Fatalf("writing package: %s", err)
+	}
+
+	out, err := cmdTestwrapper(t, testfile, "-race").CombinedOutput()
+	if code, ok := errExitCode(err); !ok || code != 1 {
+		t.Fatalf("testwrapper %s -race: expected exit code 1, got %v; output was:\n%s", testfile, err, out)
+	}
+	if want := "WARNING: DATA RACE"; !bytes.Contains(out, []byte(want)) {
+		t.Fatalf("expected race report in output, got:\n%s", out)
+	}
+	if want := "Not retrying flaky tests"; !bytes.Contains(out, []byte(want)) {
+		t.Fatalf("expected no-retry message in output, got:\n%s", out)
+	}
+	if got := bytes.Count(out, []byte("Attempt #")); got != 0 {
+		t.Fatalf("expected no retry attempts to be made, but %d were:\n%s", got, out)
+	}
+	if got := bytes.Count(out, []byte("=== RUN   TestFlaky")); got != 1 {
+		t.Fatalf("expected TestFlaky to be run exactly once, ran %d times:\n%s", got, out)
+	}
+
+	if testing.Verbose() {
+		t.Logf("success - output:\n%s", out)
+	}
+}
+
+// TestRaceAttributedToPassingTest covers the case where go test
+// attributes a data race report to a test that itself reports PASS
+// (e.g. when the racing goroutines outlive the test that spawned
+// them and TSan prints during a different test's window). Without
+// the race-detection fix, the WARNING: DATA RACE block would be
+// stuck in a passing test's log buffer and dropped on the floor.
+// See https://github.com/tailscale/tailscale/issues/19603.
+func TestRaceAttributedToPassingTest(t *testing.T) {
+	if runtime.GOOS != "linux" || runtime.GOARCH != "amd64" {
+		t.Skip("test requires the race detector, which needs linux/amd64")
+	}
+	t.Parallel()
+
+	testfile := filepath.Join(t.TempDir(), "race_test.go")
+	code := []byte(`package race_test
+
+import (
+	"sync"
+	"testing"
+)
+
+var counter int
+var wg sync.WaitGroup
+
+func TestSpawn(t *testing.T) {
+	wg.Add(2)
+	go func() { defer wg.Done(); counter++ }()
+	go func() { defer wg.Done(); counter++ }()
+}
+
+func TestWait(t *testing.T) {
+	wg.Wait()
+}
+`)
+	if err := os.WriteFile(testfile, code, 0o644); err != nil {
+		t.Fatalf("writing package: %s", err)
+	}
+
+	out, err := cmdTestwrapper(t, testfile, "-race").CombinedOutput()
+	if code, ok := errExitCode(err); !ok || code != 1 {
+		t.Fatalf("testwrapper %s -race: expected exit code 1, got %v; output was:\n%s", testfile, err, out)
+	}
+	if want := "WARNING: DATA RACE"; !bytes.Contains(out, []byte(want)) {
+		t.Fatalf("expected race report in output, got:\n%s", out)
+	}
+
+	if testing.Verbose() {
+		t.Logf("success - output:\n%s", out)
+	}
+}
+
 func TestBuildError(t *testing.T) {
 	t.Parallel()
 

